Introduce Reply Paths for BOLT12 Invoice in Offers Flow.

shaavan opened the pull request:
This PR builds on [#3087](https://github.com/lightningdevkit/rust-lightning/pull/3087) and addresses [this comment](https://github.com/lightningdevkit/rust-lightning/pull/3087#discussion_r1669513331).

### Changes:

1. Updates the Offers message flow to include `reply_path` with the sent `BOLT12Invoice`.
2. Ensures that in case of an error, the counterparty can send back any `InvoiceError` along the `reply_path`.
3. Updates the Offers test to check for the invoice's `reply_path` wherever applicable.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10364,7 +10364,7 @@ where
 
 				match response {
-					Ok(invoice) => responder.respond(OffersMessage::Invoice(invoice)),
+					Ok(invoice) => responder.respond_with_reply_path(OffersMessage::Invoice(invoice)),
```
Note that once you rebase, you'll need to add an `OffersContext` here.

shaavan replied:
Introduced in [pr3163.02](https://github.com/shaavan/rust-lightning/commits/pr3163.02)
Thanks for the pointer!

jkczyz commented on lightning/src/ln/offers_tests.rs:
```diff
@@ -512,6 +512,12 @@ fn creates_and_pays_for_offer_using_two_hop_blinded_path() {
 	create_announced_chan_between_nodes_with_value(&nodes, 2, 4, 10_000_000, 1_000_000_000);
 	create_announced_chan_between_nodes_with_value(&nodes, 2, 5, 10_000_000, 1_000_000_000);
 
+	// Create an announced channel between Bob and node[6] to ensure that in the event of a tie
+	// between Bob and Charlie for becoming the introduction node candidate, Bob is deterministically
+	// preferred over Charlie.
+	create_announced_chan_between_nodes_with_value(&nodes, 1, 6, 10_000_000, 1_000_000_000);
```
Is there a reason why this is needed here but not in other tests? Instead of doing this, could we instead call `disconnect_peers` as/when needed to avoid the non-determinism?

jkczyz replied:
I see, thanks for explaining. So it isn't failing in other cases because they are either only one hop, and thus have a smaller network or is the refund case, which has less interaction.

Anyhow, we can avoid adding another node and channel to the network by changing the assert to pass for either possible introduction node.

```rust
assert!(
	matches!(
		reply_path.introduction_node(),
		&IntroductionNode::NodeId(node_id) if node_id == bob_id || node_id == charlie_id,
	)
);
```

shaavan replied:
Definitely looks like a cleaner solution!
Updated the check, and removed the extra channel in [pr3163.07](https://github.com/shaavan/rust-lightning/commits/pr3163.07)
Thanks a lot for the suggestion!

shaavan commented:
Updated from [pr3163.01](https://github.com/shaavan/rust-lightning/commits/pr3163.01) to [pr3163.02](https://github.com/shaavan/rust-lightning/commits/pr3163.02) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3163.01..pr3163.02)):
Addressed @jkczyz and @TheBlueMatt comments

Changes:
1. Rebase on main.
2. Introduce a _fixup_ commit to introduce `context` in `respond_with_reply_path`.

jkczyz commented:
> Please include details about what you're doing and _why_ in the commit messages.

Sorry about being lax when reviewing this. @shaavan I recommend this reference, which we mention in the project contributor guidelines: https://cbea.ms/git-commit/

shaavan commented:
Updated from [pr3163.02](https://github.com/shaavan/rust-lightning/commits/pr3163.02) to [pr3163.03](https://github.com/shaavan/rust-lightning/commits/pr3163.03) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3163.02..pr3163.03)):
Addressed @TheBlueMatt and @jkczyz comments

Changes:
1. Squashed the commits
2. Update the Commit description to be clearly detailed.


Thanks, @TheBlueMatt for the suggestion, and @jkczyz for the resource!

shaavan commented:
Updated from [pr3163.03](https://github.com/shaavan/rust-lightning/commits/pr3163.03) to [pr3163.04](https://github.com/shaavan/rust-lightning/commits/pr3163.04) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3163.03..pr3163.04)):

Changes:
1. Rebase on main, to resolve merge conflicts.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10620,7 +10620,10 @@ where
 				match response {
-					Ok(invoice) => responder.respond(OffersMessage::Invoice(invoice)),
+					Ok(invoice) => {
+						let context = MessageContext::Offers(OffersContext::Unknown {});
```
Yeah, I'd figure we'd just log. What is "BP" refer to here?

jkczyz replied:
Oh, blinded path. 😛 We could though maybe a lower log level is fine for now. Not sure how much we can do about sending another invoice without another request.

shaavan commented:
Updated from [pr3163.04](https://github.com/shaavan/rust-lightning/commits/pr3163.04) to [pr3163.05](https://github.com/shaavan/rust-lightning/commits/pr3163.05) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3163.04..pr3163.05)):

Changes:
1. Moved the relevant change to the relevant commit to have a clean commit history.

shaavan commented:
Updated from [pr3163.05](https://github.com/shaavan/rust-lightning/commits/pr3163.05) to [pr3163.06](https://github.com/shaavan/rust-lightning/commits/pr3163.06) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3163.05..pr3163.06)):

Changes:
1. Rebase on main.

shaavan commented:
Updated from [pr3163.06](https://github.com/shaavan/rust-lightning/commits/pr3163.06) to [pr3163.07](https://github.com/shaavan/rust-lightning/commits/pr3163.07) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3163.06..pr3163.07)):
Addressed @jkczyz comment

1. Update the check in a test to maintain channel setup consistency with the other tests.

jkczyz commented on lightning/src/ln/offers_tests.rs:
```diff
@@ -508,15 +508,15 @@ fn creates_and_pays_for_offer_using_two_hop_blinded_path() {
-	let node_cfgs = create_node_cfgs(6, &chanmon_cfgs);
+	let chanmon_cfgs = create_chanmon_cfgs(7);
+	let node_cfgs = create_node_cfgs(7, &chanmon_cfgs);
 
 	*node_cfgs[1].override_init_features.borrow_mut() = Some(features);
 
 	let node_chanmgrs = create_node_chanmgrs(
-		6, &node_cfgs, &[None, Some(accept_forward_cfg), None, None, None, None]
+		7, &node_cfgs, &[None, Some(accept_forward_cfg), None, None, None, None, None]
 	);
-	let nodes = create_network(6, &node_cfgs, &node_chanmgrs);
+	let nodes = create_network(7, &node_cfgs, &node_chanmgrs);
```
Could you change this back to use six nodes?

shaavan replied:
Looks like I missed removing it when I removed the extra channel.
Thanks for catching it! 🌟
Fixed in [pr3163.08](https://github.com/shaavan/rust-lightning/commits/pr3163.08)

shaavan commented:
Updated from [pr3163.07](https://github.com/shaavan/rust-lightning/commits/pr3163.07) to [pr3163.08](https://github.com/shaavan/rust-lightning/commits/pr3163.08) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3163.07..pr3163.08)):
Addressed @jkczyz comment

1. Remove an extra redundant node introduced earlier in the test.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10891,7 +10891,10 @@ where
 				match response {
-					Ok(invoice) => Some((OffersMessage::Invoice(invoice), responder.respond())),
+					Ok(invoice) => {
+						let context = MessageContext::Offers(OffersContext::InboundPayment { payment_hash });
```
Ah, yeah, that's a good point. @shaavan we'll need to do something similar to what we do for `OffersContext::OutboundPayment` introduced in #3192.

shaavan replied:
Added HMAC verification for `OffersContext::InboundPayment` in [pr3163.09](https://github.com/shaavan/rust-lightning/commits/pr3163.09). Thanks so much, Matt for the suggestion, and thanks a lot Jeff for the reference!

shaavan commented:
Updated from [pr3163.08](https://github.com/shaavan/rust-lightning/commits/pr3163.08) to [pr3163.09](https://github.com/shaavan/rust-lightning/commits/pr3163.09) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3163.08..pr3163.09)):
Addressed @TheBlueMatt comment

Changes:
1. Introduce HMAC calculation and verification for OffersContext::InboundPayment.

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -366,6 +380,8 @@ impl_writeable_tlv_based_enum!(OffersContext,
 	},
 	(2, InboundPayment) => {
 		(0, payment_hash, required),
+		(1, nonce, required),
+		(2, hmac, required)
```
@TheBlueMatt Given `OffersContext::InboundPayment` are used in 0.0.124 for short-lived reply paths, are we ok with breaking serialization? Prior to this PR, they are only used in a reply path for an invoice sent for a refund. So we'd fail parsing a `InvoiceError` responses in that case.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10953,7 +10955,12 @@ where
 				let payment_hash = match context {
-					Some(OffersContext::InboundPayment { payment_hash }) => Some(payment_hash),
+					Some(OffersContext::InboundPayment { payment_hash, nonce, hmac }) => {
+						match signer::verify_payment_hash(payment_hash, hmac, nonce, expanded_key) {
+							Ok(_) => Some(payment_hash),
+							Err(_) => None,
+						}
```
Could you update these to use `pub` methods on `PaymentHash` as per #3242?

shaavan replied:
Thanks a lot for the suggestion!

I’ve been experimenting with implementing `pub` methods on `PaymentHash`, drawing some inspiration from #3242. However, it looks like this might be a bit more complicated than I initially thought.

The challenge arises because `PaymentHash` is part of the `lightning-types` crate, which is a dependency of the `lightning` crate. Plus, the `hmac_for_offer_payment` and `verify` methods need `ExpandedKey`, `HMAC`, and the `signer` struct, which brings us into a circular dependency situation.

So, we have a couple of options:

- We could move `PaymentHash` from the lightning-types crate to the lightning crate, or
- Implement a wrapper struct for `PaymentHash`.

I’d love to hear your thoughts on which direction you think we should take!

Thank you very much!

jkczyz replied:
Another option is to include a trait in `lightning` for these methods and implement it on `PaymentHash`.

shaavan replied:
That's sounds like a much cleaner idea!
Introduce the `Verification` trait in [pr3163.11](https://github.com/shaavan/rust-lightning/commits/pr3163.11)

Thank you so much for the suggestion! 🌟

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10893,7 +10893,12 @@ where
 				match response {
-					Ok(invoice) => Some((OffersMessage::Invoice(invoice), responder.respond())),
+					Ok(invoice) => {
+						let nonce = nonce.unwrap_or_else(|| Nonce::from_entropy_source(&*self.entropy_source));
```
Let's use always use a new `Nonce`. No need to reuse the one from the `Offer`.

shaavan commented:
Updated from [pr3163.09](https://github.com/shaavan/rust-lightning/commits/pr3163.09) to [pr3163.10](https://github.com/shaavan/rust-lightning/commits/pr3163.10) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3163.09..pr3163.10)):
Addressed @jkczyz comments

Changes:
1. Update docs.
2. Always use new nonce for the reply path sent with `Bolt12Invoice`.

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -347,6 +347,20 @@ pub enum OffersContext {
 		payment_hash: PaymentHash,
+
+		/// A nonce used for authenticating that a [`Bolt12Invoice`] is for a valid [`Refund`] or
+		/// [`InvoiceRequest`] and for deriving their signing keys.
```
Done! [pr3163.10](https://github.com/shaavan/rust-lightning/commits/pr3163.10)
Thanks for the pointer!

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -347,6 +347,20 @@ pub enum OffersContext {
+		nonce: Nonce,
+
+		/// Authentication code for the [`PaymentId`], which should be checked when the context is
+		/// used with an [`InvoiceError`].
```
Done! [pr3163.10](https://github.com/shaavan/rust-lightning/commits/pr3163.10) 🌟

shaavan commented:
Updated from [pr3163.10](https://github.com/shaavan/rust-lightning/commits/pr3163.10) to [pr3163.11](https://github.com/shaavan/rust-lightning/commits/pr3163.11) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3163.10..pr3163.11)):
Addressed @jkczyz comment

Changes:
1. Introduce `Verification` trait that define the public function required for creating and verifying HMAC.
2. Define the trait impl for `PaymentId`, and `PaymentHash`.

shaavan commented:
Updated from [pr3163.10](https://github.com/shaavan/rust-lightning/commits/pr3163.10) to [pr3163.11](https://github.com/shaavan/rust-lightning/commits/pr3163.11) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3163.10..pr3163.11)):
Addressed @jkczyz comment

Changes:
1. Introduce `Verification` trait that define the public function required for creating and verifying HMAC.
2. Define the trait impl for `PaymentId`, and `PaymentHash`.

shaavan commented:
Updated from [pr3163.11](https://github.com/shaavan/rust-lightning/commits/pr3163.11) to [pr3163.12](https://github.com/shaavan/rust-lightning/commits/pr3163.12) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3163.11..pr3163.12)):
Addressed @TheBlueMatt comment

Changes:
1. Squash fixups.
